File: C++/RestoreIPAddresses.cc
// Given a string containing only digits, restore it by returning all possible
// valid IP address combinations.

// For example:
// Given "25525511135",

// return ["255.255.11.135", "255.255.111.35"]. (Order does not matter)

class Solution {
    void generate(vector<string> &output, string cur, string &s, int beg, int n) {
        if (beg == s.size() && 4 == n) {
            cur.pop_back();
            output.push_back(cur);
        }
        if (beg + (4 - n) * 3 < s.size() || beg == s.size() || n >= 4) {
            return;
        }
        // 1 digit
        generate(output, cur + s[beg] + '.', s, beg + 1, n + 1);
        // can't have leading 0
        if (s[beg] == '0') {
            return;
        }
        // 2 digits
        if (s.size() >= beg + 2) {
            generate(output, cur + s.substr(beg, 2) + '.', s, beg + 2, n + 1);
        }
        // 3 digits, should be less than 256
        if (s.size() >= beg + 3) {
            string seg = s.substr(beg, 3);
            if (stoi(seg) < 256) {
                generate(output, cur + seg + '.', s, beg + 3, n + 1);
            }
        }
    }
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> output;
        if (s.size() < 4 || s.size() > 12) {
            return output;
        }
        generate(output, "", s, 0, 0);
        return output;
    }
};

File: Java/RestoreIPAddresses.java
// Given a string containing only digits, restore it by returning all possible valid IP address combinations.

// For example:
// Given "25525511135",

// return ["255.255.11.135", "255.255.111.35"]. (Order does not matter)

public class RestoreIPAddresses {
    public List<String> restoreIpAddresses(String s) {
        List<String> out = new ArrayList<>();
        restoreIP(s, 0, 4, new String(), out);
        return out;
    }

    void restoreIP(String s, int i, int c, String cur, List<String> out) {
        if (c == 0) {
            if (i == s.length()) {
                out.add(cur);
            }
        }
        for (int j = i+1; j <= Math.min(i+3, s.length()); j++) {
            if (s.charAt(i) == '0' && j > i+1) {
                break;
            }
            String t = s.substring(i, j);
            if (Integer.parseInt(t) < 256) {
                restoreIP(s, j, c-1, cur + t + (c > 1 ? "." : ""), out);
            }
        }
    }
}
